pkg/validation/usage_groups.go

```diff
@@ -0,0 +1,160 @@
+			Namespace: "pyroscope",
+			Name:      "usage_group_received_decompressed_total",
+			Help:      "The total number of decompressed bytes per profile received by usage group.",
+		},
+		[]string{"type", "tenant", "usage_group"},
+	)
+
+	// This is a duplicate of discarded_bytes_total, but with usage_group as a
+	// label.
+	UsageGroupDiscardedBytes = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "pyroscope",
+			Name:      "usage_group_discarded_bytes_total",
+			Help:      "The total number of bytes that were discarded by usage group.",
+		},
+		[]string{"reason", "tenant", "usage_group"},
+	)
+)
+
+type usageGroupEntry struct {
+	Name     string
+	Matchers []*labels.Matcher
+}
+
+// UsageGroupConfig is an allowlist of service names that have per-app usage
+// enabled. This allowlist is constructed on a per-tenant basis.
+type UsageGroupConfig struct {
+	tenantID string
+	config   []usageGroupEntry
+}
+
+// GetUsageGroupName matches the label set to a usage group. If no usage group
+// is matched, the default group name is used.
+func (u *UsageGroupConfig) GetUsageGroup(lbls phlaremodel.Labels) UsageGroup {
+	group := UsageGroup{
+		tenantID: u.tenantID,
+		name:     "other",
+	}
+
+	for _, cfgGroup := range u.config {
+		if matchesAll(cfgGroup.Matchers, lbls) {
+			group.name = cfgGroup.Name
+		}
+	}
+	return group
+}
+
+type UsageGroup struct {
+	tenantID string
+	name     string
+}
+
+func (u UsageGroup) CountReceivedBytes(profileType string, n int64) {
+	usageGroupReceivedDecompressedBytes.WithLabelValues(profileType, u.tenantID, u.name).Add(float64(n))
+}
+
+func (u UsageGroup) CountDiscardedBytes(reason string, n int64) {
+	UsageGroupDiscardedBytes.WithLabelValues(reason, u.tenantID, u.name).Add(float64(n))
+}
+
+// DistributorUsageGroups returns the usage groups that are enabled for this
+// tenant.
+func (o *Overrides) DistributorUsageGroups(tenantID string) (*UsageGroupConfig, error) {
+	ug := &UsageGroupConfig{
+		tenantID: tenantID,
+	}
+
+	groups := o.getOverridesForTenant(tenantID).DistributorUsageGroups
+	if len(groups) == 0 {
+		return ug, nil
+	}
+
+	if len(groups) > maxUsageGroups {
+		return nil, fmt.Errorf("too many usage groups configured for tenant %q: got %d, max %d", tenantID, len(groups), maxUsageGroups)
+	}
+
+	existingNames := make(map[string]struct{}, len(groups))
+	ug.config = make([]usageGroupEntry, 0, len(groups))
+
+	for _, group := range groups {
+		for name, matchersString := range group {
+			if _, ok := existingNames[name]; ok {
+				return nil, fmt.Errorf("duplicate usage group name %q for tenant %q", name, tenantID)
+			}
+			existingNames[name] = struct{}{}
+
+			if name == "" {
+				return nil, fmt.Errorf("empty service name in usage group for tenant %q", tenantID)
+			}
+
+			if matchersString == "" {
+				return nil, fmt.Errorf("no matchers for usage group %q and tenant %q", name, tenantID)
+			}
+
+			amMatchers, err := amlabels.ParseMatchers(matchersString)
```

Comment:
Maybe I am missing something, but we probably should parse those the same way we do in the query path: https://github.com/simonswine/pyroscope/blob/ed0b5643f48792ac988d975e17e285b807d5c1f9/pkg/phlaredb/head.go#L440

Reply:
Good point, a lot of this file was copied from how Mimir implemented this. So no, you aren't missing anything, I just happened to use Mimir's parse metrics approach instead of our own. I'll switch this.